### ISSUE TITLE
Fallback Provider not switching when geth stops syncing

### ISSUE THREAD

**Issue by mohamedmansour** [prysmaticlabs/prysm#8223]
# 🐞 Bug Report

### Description

Geth is still synching and beacon switched its endpoint from fallback to primary, it should stay in primary until geth node is fully sync.

### Has this worked before in a previous version?

No idea

## 🔬 Minimal Reproduction

- Mess up your clock (I do this in a WSL) to be behind
- Start GETH it is trying to sync
- Start Beacon, it will complain:
  ```
  ERROR powchain: Unable to process past headers provided time is later than the current eth1 head
  ```
- Fix time by syncing.
- Restart Beacon

## 🔥 Error

When starting up geth (still synching with right time) this is correct
```
[2021-01-06 22:41:59]  INFO powchain: Connected to eth1 proof-of-work chain endpoint=/home/mmansour/.ethereum/goerli/geth.ipc
[2021-01-06 22:42:02] ERROR powchain: Unable to process past headers provided time is later than the current eth1 head
```
Fixed time and still getting the error, which is correct because geth node is in a weird state from time, it doesn't recover.
```
[2021-01-06 22:43:17]  INFO powchain: Connected to eth1 proof-of-work chain endpoint=/home/mmansour/.ethereum/goerli/geth.ipc
[2021-01-06 22:43:17] ERROR powchain: Unable to process past headers provided time is later than the current eth1 head
```
Kill Geth, which properly does a fallback:
```
[2021-01-06 22:43:47] ERROR powchain: Could not connect to powchain endpoint error=dial unix /home/mmansour/.ethereum/goerli/geth.ipc: connect: connection refused
could not dial eth1 nodes
github.com/prysmaticlabs/prysm/beacon-chain/powchain.(*Service).connectToPowChain
        beacon-chain/powchain/service.go:384
github.com/prysmaticlabs/prysm/beacon-chain/powchain.(*Service).waitForConnection
        beacon-chain/powchain/service.go:468
github.com/prysmaticlabs/prysm/beacon-chain/powchain.(*Service).retryETH1Node
        beacon-chain/powchain/service.go:554
github.com/prysmaticlabs/prysm/beacon-chain/powchain.(*Service).initPOWService
        beacon-chain/powchain/service.go:737
github.com/prysmaticlabs/prysm/beacon-chain/powchain.(*Service).run
        beacon-chain/powchain/service.go:749
github.com/prysmaticlabs/prysm/beacon-chain/powchain.(*Service).Start.func1
        beacon-chain/powchain/service.go:275
runtime.goexit
        GOROOT/src/runtime/asm_amd64.s:1374
[2021-01-06 22:44:02]  INFO powchain: Falling back to alternative endpoint: https://goerli.infura.io/v3/[id]
[2021-01-06 22:44:18]  INFO powchain: Connected to eth1 proof-of-work chain endpoint=https://goerli.infura.io/v3/[id]
```
Now Launch Geth again, it automatically switched back to geth, but geth is not ready, its not synced yet, this shows geth is not synced.
```
INFO [01-06|22:44:23.824] Looking for peers                        peercount=0 tried=22 static=0
INFO [01-06|22:51:09.250] Looking for peers                        peercount=1 tried=8  static=0
INFO [01-06|22:57:21.716] Imported new chain segment               blocks=856  txs=2828 mgas=952.467 elapsed=8.028s    mgasps=118.629 number=4040697 hash="7524ad…30a411" age=3d16h41m dirty=255.90MiB
INFO [01-06|22:51:26.487] Imported new chain segment               blocks=209  txs=765 mgas=246.892 elapsed=8.129s mgasps=30.368 number=4014624 hash="3fec86…9b8a46" age=1w1d5h   dirty=15.73MiB
```
This shows beacon 
```
[2021-01-06 22:44:28]  INFO powchain: Primary endpoint ready again, switching back to it
[2021-01-06 22:44:43]  INFO powchain: Connected to eth1 proof-of-work chain endpoint=/home/mmansour/.ethereum/goerli/geth.ipc
[2021-01-06 22:44:43]  WARN powchain: eth1 client is not syncing
```

## 🌍  Your Environment

**Operating System:**

WSL (I use it to simulate time sync issues)

**What version of Prysm are you running? (Which release)**

<pre>
  <code>
0e5da504f46a7f0b56fd360c270d0f2f3b05e07d
  </code>
</pre>

**Comment by mohamedmansour** [comment]
I think I found an issue, I tried again, this time just started Geth and then Started Beacon:

Geth had a sync issue and it failed, it dropped my peers and trying to find new peers.
```
INFO [01-06|23:25:00.609] Imported new chain segment               blocks=461  txs=1462 mgas=598.726  elapsed=4.636s    mgasps=129.126 number=4061980 hash="6cb1d2…322a62" age=27m25s    dirty=255.25MiB
INFO [01-06|23:25:00.609] Downloader queue stats                   receiptTasks=0 blockTasks=0  itemSize=2.57KiB throttle=8192
INFO [01-06|23:25:01.597] Imported new chain segment               blocks=92   txs=238  mgas=111.536  elapsed=987.027ms mgasps=113.002 number=4062072 hash="a9e220…5a02c5" age=4m26s     dirty=254.74MiB
WARN [01-06|23:25:01.597] Synchronisation failed, dropping peer    peer=1de6fc6c000adb83 err="action from bad peer ignored: stale headers"
INFO [01-06|23:25:04.328] Looking for peers                        peercount=0 tried=4  static=0
INFO [01-06|23:25:14.330] Looking for peers                        peercount=0 tried=4  static=0
INFO [01-06|23:25:21.764] Deep froze chain segment                 blocks=86   elapsed=63.902ms  number=3972072 hash="fb8f63…f51b52"
INFO [01-06|23:25:24.332] Looking for peers                        peercount=0 tried=4  static=0
INFO [01-06|23:25:34.333] Looking for peers                        peercount=0 tried=3  static=0
```

Beacon instantly switched back to GETH, which it shouldn't
```
[2021-01-06 23:24:59]  INFO blockchain: Finished applying state transition attestations=85 attesterSlashings=0 deposits=0 proposerSlashings=0 voluntaryExits=0
[2021-01-06 23:25:07]  INFO blockchain: Synced new block block=0x83f02dfc... epoch=11207 finalizedEpoch=11205 finalizedRoot=0xd71e85fa... slot=358625 slotInEpoch=1
[2021-01-06 23:25:07]  INFO blockchain: Finished applying state transition attestations=73 attesterSlashings=0 deposits=0 proposerSlashings=0 voluntaryExits=0
[2021-01-06 23:25:08]  INFO powchain: Primary endpoint ready again, switching back to it
[2021-01-06 23:25:23]  INFO powchain: Connected to eth1 proof-of-work chain endpoint=/home/mmansour/.ethereum/goerli/geth.ipc
```

**Comment by terencechain** [comment]
cc @nisdas

**Comment by mohamedmansour** [comment]
Geth is waiting for peers to sync (so its not syncing yet)
```
INFO [01-10|11:23:49.648] Looking for peers                        peercount=0 tried=4 static=0
INFO [01-10|11:23:59.649] Looking for peers                        peercount=0 tried=4 static=0
INFO [01-10|11:24:09.650] Looking for peers                        peercount=0 tried=4 static=0
```

But beacon still thinks it is working properly:
```
[2021-01-10 11:24:47]  INFO powchain: Connected to eth1 proof-of-work chain endpoint=/home/mmansour/.ethereum/goerli/geth.ipc
[2021-01-10 11:24:47] ERROR powchain: Unable to process past headers provided time is later than the current eth1 head
```

Where it should fallback to the infura node, but it isn't.

What we can do here, if the ETH1 endpoint has no peers (`admin.peers.length`) then we can assume its not ready and we should fallback to the fallback provider (infura/etc). Is that reasonable @nisdas ?

**Comment by nisdas** [comment]
It wouldn't be difficult to do @mohamedmansour , the main issue is that most beacon node's do not have elevated access to the eth1 node. So performing this sort of request would only work in a few cases(beacon node has admin access). Also that particular method is a non-standard rpc method, so it would only work for geth.

**Comment by mohamedmansour** [comment]
@nisdas perhaps another approach would be keeping track of the eth1 head block for the health check. If one becomes stale we can safely assume its unhealthy, and the fallback would be promoted.

**Comment by handelaar2** [comment]
Perhaps yet another approach would be to 'just' wait an extra 15-30 minutes before doing the actual switch back to primary after it comes back.

**Comment by DanielFontes** [comment]
I'm having the same problem today because of the #12244294 bug in OpenEthereum. Prysm reports that "eth1 client is not syncing" but doesn't switch to the fallbacks.

**Comment by nisdas** [comment]
@DanielFontes This will be fixed in the next release, I have a PR up: #8771 .
